### PR TITLE
Add programming guide for error handling

### PR DESCRIPTION
**Pull request by jbartok**
Needed as a discoverable place about the whole error handling topic.

### CODE REVIEW

File: site/docs/api/error-handling.md
Comment (nfrankel) [inline site/docs/api/error-handling.md:110]: Learned a new word: _resumption_

File: site/docs/api/error-handling.md
Comment (nfrankel) [inline site/docs/api/error-handling.md:35]: Why use Jet here? We can also use an unreachable web-service.

**Comment by nfrankel** [comment]
Great idea! I was thinking about that the other day, and I was missing this data.

1. Perhaps some code snippets within specific use-cases would be a good addition?
2. I'm a visual person. Perhaps add a diagram to make it more visual? A simple summary of IF a, THEN A, IF b, THEN b

File: site/docs/api/error-handling.md
Comment (vladoschreiner) [inline site/docs/api/error-handling.md:97]: So why do the jobs snapshot? It feels that there is no benefit unless one configures suspend on failure

It's important to say that snapshots are useful to overcome member failures.

File: site/docs/api/error-handling.md
Comment (vladoschreiner) [inline site/docs/api/error-handling.md:136]: Catch exception on a processor level as another section. This is what we recommend to deal with user code.

**Review by vladoschreiner — CHANGES_REQUESTED** [review]

File: site/docs/api/error-handling.md
Comment (viliam-durina) [inline site/docs/api/error-handling.md]: This discrimination is too anecdotal, we can do something more abstract:
```suggestion
* *input data error*: unexpected input data, e.g. unexpected null value, out-of-range date (`0000-00-00`) or malformed JSON document
* *coding errors*: e.g. NullPointerException, memory leaks, ...
* *I/O errors*: network errors, remote service failures
```
Also "ex." -> "for example," or "e.g."

File: site/docs/api/error-handling.md
Comment (viliam-durina) [inline site/docs/api/error-handling.md:47]: "This are" -> "These are"

File: site/docs/api/error-handling.md
Comment (viliam-durina) [inline site/docs/api/error-handling.md:51]: "but not just *those*"
"to Jet" -> "by Jet"

File: site/docs/api/error-handling.md
Comment (viliam-durina) [inline site/docs/api/error-handling.md:50]: "all error handling" -> _all errors_, really?

File: site/docs/api/error-handling.md
Comment (viliam-durina) [inline site/docs/api/error-handling.md:95]: ~their~ fault tolerance

File: site/docs/api/error-handling.md
Comment (viliam-durina) [inline site/docs/api/error-handling.md]: "this configuration option is not the default" -> "this option is disabled by default"

File: site/docs/api/error-handling.md
Comment (viliam-durina) [inline site/docs/api/error-handling.md]:
```suggestion
resumptions possible once the root cause of the failure has been
```

File: site/docs/api/error-handling.md
Comment (viliam-durina) [inline site/docs/api/error-handling.md]: It preserves exactly-once guarantee, but it's exactly-once semantics, not exactly-once processing. Exactly-once processing is impossible.

File: site/docs/api/error-handling.md
Comment (viliam-durina) [inline site/docs/api/error-handling.md:136]: We should not refer to description of important features in TDD. We can refer to TDD to read more why the design is as it is or so.

**Review by viliam-durina — COMMENTED** [review]
I don't like the style very much. It's too much prose and quite imprecise, quick in generalizations and anecdotal. But I'll approve after my comments are addressed.

File: site/docs/api/error-handling.md
Comment (jbartok) [reply site/docs/api/error-handling.md:51]: I don't agree with the "to -> by" change. I wanted to say that Jet is not aware of the error, and for that "transparent to Jet" seems the right expression.

Comment (mtopolnik) [reply site/docs/api/error-handling.md:51]: Instead of this confusion just use the active voice and say explicitly who is the actor.

Comment (viliam-durina) [reply site/docs/api/error-handling.md:51]: @jbartok you're right, I missed it.

File: site/docs/api/error-handling.md
Comment (jbartok) [reply site/docs/api/error-handling.md:136]: I don't really understand this particular comment. Our processors are not capable of dealing with their own errors right now and we've decided to not implement that at the moment, because it's not something achievable with reasonable effort. So if we would recommend anything to users, that would be to catch exceptions from their own code, which is just common sense, I think. 

Did you mean something else?

File: site/docs/api/error-handling.md
Comment (jbartok) [reply site/docs/api/error-handling.md]: You are perfectly right, theoretically. Batch jobs do have state but they don't actually save it, so in practical terms they don't have state... You shouldn't even enable snapshotting for batch jobs, because there are known scenarios when it wouldn't work well (for example batch job with keyed windowing). Viliam can list a ton of related issues just from the top of his head... I will try to find the right formulation.

File: site/docs/api/error-handling.md
Comment (jbartok) [reply site/docs/api/error-handling.md:96]: I will emphasize that.

**Comment by jbartok** [comment]
I've tried to address all concerns expressed so far. Pls. give it another round of comments. I can't say that I've enjoyed them so far, but the results seems (and hopefully is) better.

File: site/docs/api/error-handling.md
Comment (viliam-durina) [inline site/docs/api/error-handling.md:10]: - scheduled maintenance isn't an error condition: you can do an orderly shutdown

**Review by viliam-durina — APPROVED** [review]

**Comment by jbartok** [comment]
Thanks @viliam-durina, I'll wait for @vladoschreiner 's approval too.

File: site/docs/api/error-handling.md
Comment (vladoschreiner) [reply site/docs/api/error-handling.md]: > if we would recommend anything to users, that would be to catch exceptions from their own code, which is just common sense

That is what I thought. We can skip this part for now.

**Review by vladoschreiner — APPROVED** [review]

**Pull request merged**
